base/main_test_bucket_pool.go

```diff
@@ -0,0 +1,650 @@
+				atomic.AddInt64(&tbp.stats.TotalBucketReadierDurationNano, time.Since(start).Nanoseconds())
+			}(testBucketName)
+		}
+	}
+
+	tbp.Logf(context.Background(), "Stopped bucketReadier")
+}
+
+// openTestBucket opens the bucket of the given name for the gocb cluster in the given GocbTestBucketPool.
+func (tbp *GocbTestBucketPool) openTestBucket(testBucketName tbpBucketName, sleeper RetrySleeper) (*CouchbaseBucketGoCB, error) {
+
+	ctx := bucketNameCtx(context.Background(), string(testBucketName))
+
+	bucketSpec := tbp.defaultBucketSpec
+	bucketSpec.BucketName = string(testBucketName)
+
+	waitForNewBucketWorker := func() (shouldRetry bool, err error, value interface{}) {
+		gocbBucket, err := GetCouchbaseBucketGoCBFromAuthenticatedCluster(tbp.cluster, bucketSpec, "")
+		if err != nil {
+			tbp.Logf(ctx, "Retrying OpenBucket")
+			return true, err, nil
+		}
+		return false, nil, gocbBucket
+	}
+
+	tbp.Logf(ctx, "Opening bucket")
+	err, val := RetryLoop("waitForNewBucket", waitForNewBucketWorker, sleeper)
+
+	gocbBucket, _ := val.(*CouchbaseBucketGoCB)
+
+	return gocbBucket, err
+}
+
+// TBPBucketInitFunc is a function that is run once (synchronously) when creating/opening a bucket.
+type TBPBucketInitFunc func(ctx context.Context, b Bucket, tbp *GocbTestBucketPool) error
+
+// NoopInitFunc does nothing to init a bucket. This can be used in conjunction with FlushBucketReadier when there's no requirement for views/GSI.
+var NoopInitFunc TBPBucketInitFunc = func(ctx context.Context, b Bucket, tbp *GocbTestBucketPool) error {
+	return nil
+}
+
+// PrimaryIndexInitFunc creates a primary index on the given bucket. This can then be used with N1QLBucketEmptierFunc, for improved compatibility with GSI.
+var PrimaryIndexInitFunc TBPBucketInitFunc = func(ctx context.Context, b Bucket, tbp *GocbTestBucketPool) error {
+	gocbBucket, ok := AsGoCBBucket(b)
+	if !ok {
+		tbp.Logf(ctx, "skipping primary index creation for non-gocb bucket")
+		return nil
+	}
+
+	if hasPrimary, _, err := gocbBucket.getIndexMetaWithoutRetry(PrimaryIndexName); err != nil {
+		return err
+	} else if !hasPrimary {
+		err := gocbBucket.CreatePrimaryIndex(PrimaryIndexName, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TBPBucketReadierFunc is a function that runs once a test is finished with a bucket. This runs asynchronously.
+type TBPBucketReadierFunc func(ctx context.Context, b *CouchbaseBucketGoCB, tbp *GocbTestBucketPool) error
+
+// FlushBucketEmptierFunc ensures the bucket is empty by flushing. It is not recommended to use with GSI.
+var FlushBucketEmptierFunc TBPBucketReadierFunc = func(ctx context.Context, b *CouchbaseBucketGoCB, tbp *GocbTestBucketPool) error {
+	return b.Flush()
+}
+
+// N1QLBucketEmptierFunc ensures the bucket is empty by using N1QL deletes. This is the preferred approach when using GSI.
+var N1QLBucketEmptierFunc TBPBucketReadierFunc = func(ctx context.Context, b *CouchbaseBucketGoCB, tbp *GocbTestBucketPool) error {
```

Comment:
It doesn't look like N1QLBucketEmptierFunc is being used as a TBPBucketReadierFunc - it's just being invoked directly from inside ViewsAndGSIBucketReadier.  Should this be switched to a normal utility function, or do you anticipate future use cases as a TBPBucketReadier?

Reply:
Required when re-enabling GSI. Updated with comment.